jax/_src/scipy/special.py

```diff
@@ -1010,3 +1011,94 @@ def lpmn_values(m: int, n: int, z: jnp.ndarray, is_normalized: bool) -> jnp.ndar
   l_max = n
 
   return _gen_associated_legendre(l_max, z, is_normalized)
+
+
+
+@partial(jit, static_argnums=(4,))
+def _sph_harm(m: jnp.ndarray,
+              n: jnp.ndarray,
+              theta: jnp.ndarray,
+              phi: jnp.ndarray,
+              n_max: int) -> jnp.ndarray:
+  """Computes the spherical harmonics."""
+
+  len_inputs = m.shape[0]
+  harmonics = jnp.zeros((len_inputs), dtype=jnp.complex64)
+  cos_colatitude = jnp.cos(phi)
+
+  legendre = _gen_associated_legendre(n_max, cos_colatitude, True)
+
+  def body_fun(idx, harmonics):
+    degree = n[idx]
+    order = m[idx]
+    nonnegative_order = jnp.abs(order)
+
+    legendre_val = legendre[nonnegative_order, degree, idx]
+
+    # Phase term exp(1j * nonnegative_order * longitude).
+    longitude = theta[idx]
+    angle = nonnegative_order * longitude
+    vandermonde = lax.complex(jnp.cos(angle), jnp.sin(angle))
+
+    harm_iter = lax.complex(legendre_val * jnp.real(vandermonde),
+                            legendre_val * jnp.imag(vandermonde))
+
+    # Negative order.
+    harm_iter = jnp.where(order < 0,
+                          (-1.0)**abs(order) * jnp.conjugate(harm_iter),
+                          harm_iter)
+
+    harmonics = harmonics.at[idx].set(harm_iter)
+
+    return harmonics
+
+  harmonics = lax.fori_loop(
+      lower=0, upper=len_inputs, body_fun=body_fun, init_val=harmonics)
```

Comment:
Looking at this, I think you may be able to avoid the `lax.fori_loop` in favor of faster vectorized computations – that is, something like this (untested):
```python
legendre_val = legendre[abs(m), n, jnp.arange(len(n))]
angle = abs(m) * theta
vandermonde = lax.complex(jnp.cos(angle), jnp.sin(angle))
harmonics = lax.complex(legendre_val * jnp.real(vandermonde),
                        legendre_val * jnp.imag(vandermonde))
# Negative order.
harmonics = jnp.where(m < 0,
                      (-1.0)**abs(m) * jnp.conjugate(harmonics),
                      harmonics)
```

Reply:
Brilliant idea, Jake! BTW, it passed the test without a single edit.